models/baseModels/tests/testLoyaltyProgram.spec.ts

```diff
@@ -0,0 +1,275 @@
+    `Loyalty Program '${loyaltyProgramData.name}' exists `
+  );
+
+  const partyDoc = (await fyo.doc.getDoc(
+    ModelNameEnum.Party,
+    partyData.name
+  )) as Party;
+
+  partyDoc.loyaltyProgram = loyaltyProgramData.name;
+
+  await partyDoc.sync();
+
+  t.equals(
+    partyDoc.loyaltyProgram,
+    loyaltyProgramData.name,
+    `Loyalty Program '${loyaltyProgramData.name}' successfully added to Party '${partyData.name}'`
+  );
+});
+
+async function loyaltyPointEntryDoc(sinvName: string) {
+  const loyaltyPointEntryData = (await fyo.db.getAll(
+    ModelNameEnum.LoyaltyPointEntry,
+    {
+      fields: ['name', 'customer', 'loyaltyPoints', 'loyaltyProgramTier'],
+      filters: { invoice: sinvName! },
+    }
+  )) as {
+    name?: string;
+    customer?: string;
+    loyaltyPoints?: number;
+    loyaltyProgramTier?: string;
+  }[];
+
+  if (loyaltyPointEntryData) {
+    return loyaltyPointEntryData[0];
+  }
+}
+
+async function createSalesInvoice() {
+  const sinvDoc = fyo.doc.getNewDoc(ModelNameEnum.SalesInvoice, {
+    account: 'Debtors',
+    party: partyData.name,
+    items: [
+      {
+        item: itemData.name,
+        rate: itemData.rate,
+        quantity: 1,
+      },
+    ],
+  }) as SalesInvoice;
+
+  return sinvDoc;
+}
+
+test('create Sales Invoice and verify loyalty points are created correctly', async (t) => {
+  const sinvDoc = await createSalesInvoice();
+
+  await sinvDoc.sync();
+  await sinvDoc.runFormulas();
+  await sinvDoc.submit();
+
+  t.ok(
+    fyo.db.exists(ModelNameEnum.SalesInvoice, sinvDoc.name),
+    `Sales Invoice '${sinvDoc.name}' exists`
+  );
+
+  t.ok(
+    fyo.db.exists(ModelNameEnum.SalesInvoice, sinvDoc.loyaltyProgram),
+    `Loyalty Program '${sinvDoc.loyaltyProgram}' should be linked to Sales Invoice '${sinvDoc.name}'`
+  );
+
+  const loyaltyPointEntryData = await loyaltyPointEntryDoc(
+    sinvDoc.name as string
+  );
+
+  t.ok(
+    await fyo.db.exists(
+      ModelNameEnum.LoyaltyPointEntry,
+      loyaltyPointEntryData?.name
+    ),
+    `Loyalty Point Entry '${loyaltyPointEntryData?.name}' should be created against Sales Invoice '${sinvDoc.name}'`
+  );
+
+  const loyaltyProgramDoc = (await fyo.doc.getDoc(
+    ModelNameEnum.LoyaltyProgram,
+    sinvDoc.loyaltyProgram
+  )) as Party;
+
+  const selectedTier: CollectionRulesItems | undefined = getLoyaltyProgramTier(loyaltyProgramDoc, fyo.pesa(itemData.rate));
+
+  t.equals(
+    loyaltyPointEntryData?.loyaltyProgramTier,
+    selectedTier?.tierName,
+    `tier must be ${loyaltyPointEntryData?.loyaltyProgramTier} because it is the matched tier'`
+  );
+
+  const tierData: any = collectionRulesData.find((rule) => {
+    return rule.tierName === loyaltyPointEntryData?.loyaltyProgramTier;
+  });
+
+  t.equals(
+    loyaltyPointEntryData?.loyaltyPoints,
+    (itemData.rate * tierData?.collectionFactor) as number,
+    `${loyaltyPointEntryData?.loyaltyPoints} loyalty points should be added to customer '${loyaltyPointEntryData?.customer}'`
+  );
```

Comment:
message doesn't explain the testcase.

should be something like _Loyalty Point calculation is correct_